RetailCoder.VBE/Inspections/VariableRequiresSetAssignmentEvaluator.cs

```diff
@@ -0,0 +1,160 @@
+        public VariableRequiresSetAssignmentEvaluator(RubberduckParserState parserState)
+        {
+            _parserState = parserState;
+        }
+
+        public IEnumerable<Declaration> GetDeclarationsPotentiallyRequiringSetAssignment()
+        {
+            var interestingDeclarations = _parserState.AllUserDeclarations.Where(item =>
+                    IsVariableOrParameter(item)
+                    && !item.IsSelfAssigned
+                    && TypeIsAnObjectOrVariant(item));
+
+            var interestingMembers = _parserState.AllUserDeclarations.Where(item =>
+                    IsMemberWithReturnType(item)
+                    && item.IsTypeSpecified
+                    && TypeIsAnObjectOrVariant(item));
+
+            var allInterestingDeclarations = interestingDeclarations
+                    .Union(HasReturnAssignment(interestingMembers));
+
+            return allInterestingDeclarations;
+        }
+
+        public bool RequiresSetAssignment(IdentifierReference reference)
+        {
+            var declaration = reference.Declaration;
+            var MayRequireAssignmentUsingSet =
+                 (IsVariableOrParameter(declaration) || IsMemberWithReturnType(declaration) )
+                 && !declaration.IsSelfAssigned
+                 && TypeIsAnObjectOrVariant(declaration);
+
+            if(!MayRequireAssignmentUsingSet) { return false; }
+
+            var allInterestingDeclarations = GetDeclarationsPotentiallyRequiringSetAssignment();
+
+            return ObjectOrVariantRequiresSetAssignment(reference, allInterestingDeclarations);
+        }
+
+        private bool IsMemberWithReturnType(Declaration item)
+        {
+            return (item.DeclarationType == DeclarationType.Function
+                || item.DeclarationType == DeclarationType.PropertyGet);
+        }
+
+        private IEnumerable<Declaration> HasReturnAssignment(IEnumerable<Declaration> interestingMembers)
+        {
+            return interestingMembers.SelectMany(member =>
+                      member.References.Where(memberRef => memberRef.ParentScoping.Equals(member)
+                           && memberRef.IsAssignment)).Select(reference => reference.Declaration);
+        }
+
+        private bool IsVariableOrParameter(Declaration item)
+        {
+            return item.DeclarationType == DeclarationType.Variable
+                    || item.DeclarationType == DeclarationType.Parameter;
+        }
+
+        private bool TypeIsAnObjectOrVariant(Declaration item)
+        {
+            return !item.IsArray
+                    && !ValueOnlyTypes().Contains(item.AsTypeName)
+                    && (item.AsTypeDeclaration == null
+                        || TypeRequiresSetAssignment(item));
+        }
+
+        private IEnumerable<string> ValueOnlyTypes()
+        {
+            var nonSetTypes = SymbolList.ValueTypes.ToList();
+            nonSetTypes.Remove(Tokens.Variant);
```

Comment:
This should be a `.Where` - you're materializing a `List` and then iterating over it to do the `.Remove`.